proposals/extend_pod_security_admission.md

```diff
@@ -0,0 +1,229 @@
+  policy: require-labels
+  resources:
+  - apiVersion: v1
+    kind: Pod
+    name: nginx
+    namespace: default
+    uid: 13fcb726-5597-4fcd-a708-ce8b558bd484
+  rule: check-for-labels
+  properties:
+    controlName: Capabilities
+    podSecurity: spec.containers[*].securityContext.capabilities.add
+  scored: true
+  status: fail
+```
+
+### Exceptions for container security context
+
+* Host Namespaces: `spec.hostNetwork,spec.hostPID,spec.hostIPC`
+* Sysctls: `spec.securityContext.sysctls[*].name`
+* volumes: `spec.volumes`
+* Non-root groups: `spec.securityContext.supplementalGroups[*], spec.securityContext.fsGroup`
+
+
+## Alternate Solutions Considered
+### Extend Via Annotations
+
+This approach extends the PSA via annotations for granular control.
+
+#### Step 1. Enable additional policies for a namespace
+
+One should be able to configure additional pod-security policies by adding the following annotation to the Namespace:
+
+```yaml
+# Support modes are: warn, audit, enforce.
+# The value of the annotation is comma-separated policy names.
+# Multiple annotations with different modes can be configured 
+# for the same namespace.
+# If a policy is enabled for more than one mode, 
+# the most restricted mode will take precedence.
+pod-security.kyverno.io/<mode>=<policy1,policy2>
+```
+
+Once the annotation is detected, the corresponding policies (from 
+“kyverno/policies”) will be installed to the cluster as Namespace policies, and will be applied to all pods and pod controllers. The installed Namespace policy shares the same lifecycle of the configured annotation, if the policy is removed from the annotation, or if this annotation is removed completely from the Namespace, the policy will be removed from the cluster.
+
+Note: We'll need to define what are the policies we would like to support via annotation. Once set, the policy name should be barely changed.
+
+#### Step 2. Fine-grained Policy Control
+
+We can start with label selectors to select matching resources, and extend to use annotation selectors if needed. Kyverno will look up given labels for the incoming pods and pod controllers, and apply the selected policies (by the annotation above) to the matching resources. This annotation must be configured along with `pod-security.kyverno.io/<mode>` for granular control.
+
+```yaml
+# selectors can be defined either in JSON string or Yaml format
+
+# use label selector for exact match
+pod-security.kyverno.io/label-selector: >-
+    matchLabels:
+        env: prod
+        
+# use label selector to match expressions
+pod-security.kyverno.io/label-selector: >-       
+    matchExpressions:
+    - key: env
+      operator: In
+      values:
+      - "prod"
+      - "staging"
+```
+
+## Next Steps
+
+The first implementation will enable the new type of rule to be working together with PSA. The following requirements also need to be considered while developing.
+
+* Add "warn" mode
+* Support "auto-gen"
+* Metrics support
+* CLI support
```

Comment:
Btw, we can do dry-run with PSA: `kubectl label --dry-run=server --overwrite -n <my_namespace> \
    pod-security.kubernetes.io/<mode>=<level>`

Reply:
Right, the Kyverno CLI can provide the same function.